### ISSUE TITLE
CSV export does not create CSV

### ISSUE THREAD

**Issue by tnibert** [nicolargo/glances#1309]
#### Description

Export to CSV generates empty CSV when done per instructions in --help.

```
$ python -m glances --stdout cpu --export csv --export-csv-file test.csv
cpu: {'total': 23.7, 'user': 22.7, 'system': 6.8, 'idle': 70.5, 'nice': 0.0, 'iowait': 0.0, 'irq': 0.0, 'softirq': 0.0, 'steal': 0.0, 'guest': 0.0, 'guest_nice': 0.0, 'ctx_switches': 478, 'interrupts': 141, 'soft_interrupts': 352, 'syscalls': 0, 'time_since_update': 0.10971760749816895, 'cpucore': 4}
Exception in thread Thread-8:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tim/code/glances2/glances/glances/exports/glances_csv.py", line 83, in update
    plugin, self.get_item_key(stat), item) for item in stat)
  File "/home/tim/code/glances2/glances/glances/exports/glances_csv.py", line 83, in <genexpr>
    plugin, self.get_item_key(stat), item) for item in stat)
  File "/home/tim/code/glances2/glances/glances/exports/glances_export.py", line 119, in get_item_key
    if isinstance(ret, list):
UnboundLocalError: local variable 'ret' referenced before assignment

Exception in thread Thread-10:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tim/code/glances2/glances/glances/exports/glances_csv.py", line 83, in update
    plugin, self.get_item_key(stat), item) for item in stat)
  File "/home/tim/code/glances2/glances/glances/exports/glances_csv.py", line 83, in <genexpr>
    plugin, self.get_item_key(stat), item) for item in stat)
  File "/home/tim/code/glances2/glances/glances/exports/glances_export.py", line 119, in get_item_key
    if isinstance(ret, list):
UnboundLocalError: local variable 'ret' referenced before assignment
$ cat test.csv 
$ 
```

Per instructions in https://glances.readthedocs.io/en/latest/gw/csv.html.  No CSV file is generated.

```
$ python -m glances --stdout cpu --export-csv test.csv
cpu: {'total': 23.9, 'user': 19.5, 'system': 4.9, 'idle': 75.6, 'nice': 0.0, 'iowait': 0.0, 'irq': 0.0, 'softirq': 0.0, 'steal': 0.0, 'guest': 0.0, 'guest_nice': 0.0, 'ctx_switches': 182, 'interrupts': 85, 'soft_interrupts': 304, 'syscalls': 0, 'time_since_update': 0.10525369644165039, 'cpucore': 4}
$ ls
appveyor.yml  CODE-OF-CONDUCT.md  COPYING       glances      optional-requirements.txt  requirements.txt  sonar-project.properties  unitest-all.sh      unitest-xmlrpc.py
AUTHORS       conf                docker-files  MANIFEST.in  pylint.cfg                 setup.py          tox.ini                   unitest.py
circle.yml    CONTRIBUTING.md     docs          NEWS         README.rst                 snap              uninstall.sh              unitest-restful.py
```

#### Versions

* Glances & psutil (glances -V):
$ python -m glances -V
Glances v3.0.rc5 with psutil v5.4.3
(Current repository, git commit 1d2add124852cb8e88a7e99aaab631c7103f49e4)

* Operating System (lsb_release -a):
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.1 LTS
Release:	18.04
Codename:	bionic

#### Logs

https://pastebin.com/Kc9LGsk4

**Comment by nicolargo** [comment]
Hi @tnibert ,

i can not reproduce the issue on my system:

```
# python -m glances --stdout cpu --export csv --export-csv-file /tmp/test.csv
cpu: {'softirq': 0.0, 'iowait': 3.9, 'interrupts': 1554, 'system': 15.5, 'guest': 0.0, 'soft_interrupts': 1337, 'time_since_update': 0.25975990295410156, 'idle': 35.0, 'user': 45.6, 'guest_nice': 0.0, 'irq': 0.0, 'cpucore': 4, 'syscalls': 0, 'total': 23.2, 'steal': 0.0, 'ctx_switches': 3651, 'nice': 0.0}
cpu: {'softirq': 0.0, 'iowait': 0.4, 'interrupts': 7097, 'system': 3.0, 'guest': 0.0, 'soft_interrupts': 4051, 'time_since_update': 6.592353105545044, 'idle': 88.2, 'user': 8.4, 'guest_nice': 0.0, 'irq': 0.0, 'cpucore': 4, 'syscalls': 0, 'total': 4.9, 'steal': 0.0, 'ctx_switches': 26153, 'nice': 0.0}
cpu: {'softirq': 0.0, 'iowait': 0.0, 'interrupts': 265, 'system': 11.2, 'guest': 0.0, 'soft_interrupts': 641, 'time_since_update': 0.2239229679107666, 'idle': 62.9, 'user': 25.8, 'guest_nice': 0.0, 'irq': 0.0, 'cpucore': 4, 'syscalls': 0, 'total': 4.9, 'steal': 0.0, 'ctx_switches': 1895, 'nice': 0.0}
cpu: {'softirq': 0.0, 'iowait': 0.0, 'interrupts': 677, 'system': 0.7, 'guest': 0.0, 'soft_interrupts': 1068, 'time_since_update': 2.9957029819488525, 'idle': 97.3, 'user': 2.0, 'guest_nice': 0.0, 'irq': 0.0, 'cpucore': 4, 'syscalls': 0, 'total': 5.1, 'steal': 0.0, 'ctx_switches': 1661, 'nice': 0.0}
cpu: {'softirq': 0.0, 'iowait': 0.1, 'interrupts': 593, 'system': 1.1, 'guest': 0.0, 'soft_interrupts': 1006, 'time_since_update': 3.0021979808807373, 'idle': 97.0, 'user': 1.8, 'guest_nice': 0.0, 'irq': 0.0, 'cpucore': 4, 'syscalls': 0, 'total': 2.8, 'steal': 0.0, 'ctx_switches': 2782, 'nice': 0.0}
cpu: {'softirq': 0.0, 'iowait': 0.0, 'interrupts': 556, 'system': 0.9, 'guest': 0.0, 'soft_interrupts': 964, 'time_since_update': 3.003443956375122, 'idle': 97.1, 'user': 2.0, 'guest_nice': 0.0, 'irq': 0.0, 'cpucore': 4, 'syscalls': 0, 'total': 3.0, 'steal': 0.0, 'ctx_switches': 2708, 'nice': 0.0}
# ll /tmp/test.csv 
-rw-rw-r-- 1 nicolargo nicolargo 13514 août  27 18:44 /tmp/test.csv
# tail -2 /tmp/test.csv 
2018-08-27 18:44:09,4,1.39,1.21,1.05,88.169.86.29,24,255.255.255.0,192.168.0.254,192.168.0.48,0,0.0,8082419712,0,8082419712,0,284603,2.9955780506134033,3598043,True,interface_name,0,128336,305307399,3469707,324245046,18937647,wlp2s0,2.9955780506134033,6808,True,interface_name,0,3404,911872,3404,1823744,911872,lo,2.9955780506134033,0,True,interface_name,0,0,0,0,0,0,docker0,0,0.0,0.3,2.4,7.6,0.0,80.5,9.1,cpu_number,0.0,19.5,0.0,0.0,1,0.0,0.0,0.0,5.2,0.0,70.4,24.4,cpu_number,0.0,29.6,0.0,0.0,2,0.0,0.0,0.3,6.2,0.0,80.6,12.9,cpu_number,0.0,19.4,0.0,0.0,3,0.0,0.0,0.6,5.6,0.0,74.1,19.7,cpu_number,0.0,25.9,0.0,0.0,Linux,4.4.0-133-generic,Ubuntu 16.04,XPS13-9333,64bit,Ubuntu 16.04 64bit,7,31711232,115,2.9925308227539062,32768,disk_name,sda5,0,0,0,2.9925308227539062,0,disk_name,sda2,0,0,0,2.9925308227539062,0,disk_name,sda1,7,31711232,294,2.9925308227539062,32768,disk_name,dm-0,0,0,0,2.9925308227539062,0,disk_name,dm-1,/,ext4,62387564544,mnt_point,243396149248,73.0,168621129728,/dev/mapper/ubuntu--gnome--vg-root,/var/lib/docker/aufs,ext4,62387564544,mnt_point,243396149248,73.0,168621129728,/dev/mapper/ubuntu--gnome--vg-root,4908404736,2963550208,3060883456,37.6,4908404736,1297723392,4044513280,331452416,7871954944,584507392,TP-LINK_BB28,True,-39,ssid,wpa2,70/70,Freebox-61A799,True,-85,ssid,wpa2,25/70,FreeWifi,False,-86,ssid,,24/70,FreeWifi_secure,True,-86,ssid,wpa2,24/70,FreeWifi_secure,True,-37,ssid,wpa2,70/70,freebox_NH,True,-37,ssid,wpa,70/70,FreeWifi,False,-39,ssid,,70/70,0.0,0.4,7097,3.0,0.0,4051,6.592353105545044,88.2,8.4,0.0,0.0,4,0,4.9,0.0,26153,0.0
2018-08-27 18:44:09,4,1.28,1.19,1.05,88.169.86.29,24,255.255.255.0,192.168.0.254,192.168.0.48,0,0.0,8082419712,0,8082419712,0,284607,3.8367269039154053,16301,True,interface_name,0,4832,305318868,11469,324261347,18942479,wlp2s0,3.8367269039154053,1100,True,interface_name,0,550,912422,550,1824844,912422,lo,3.8367269039154053,0,True,interface_name,0,0,0,0,0,0,docker0,0,0.0,0.0,1.6,1.8,0.0,92.4,4.2,cpu_number,0.0,7.6,0.0,0.0,1,0.0,0.0,0.0,1.6,0.0,90.6,7.9,cpu_number,0.0,9.4,0.0,0.0,2,0.0,0.0,0.0,1.6,0.0,94.8,3.7,cpu_number,0.0,5.2,0.0,0.0,3,0.0,0.0,0.0,2.1,0.0,93.7,4.2,cpu_number,0.0,6.3,0.0,0.0,Linux,4.4.0-133-generic,Ubuntu 16.04,XPS13-9333,64bit,Ubuntu 16.04 64bit,2,4104192,85,3.8352620601654053,8192,disk_name,sda5,0,0,0,3.8352620601654053,0,disk_name,sda2,0,0,0,3.8352620601654053,0,disk_name,sda1,2,4104192,270,3.8352620601654053,8192,disk_name,dm-0,0,0,0,3.8352620601654053,0,disk_name,dm-1,/,ext4,62522568704,mnt_point,243396149248,72.9,168486125568,/dev/mapper/ubuntu--gnome--vg-root,/var/lib/docker/aufs,ext4,62522568704,mnt_point,243396149248,72.9,168486125568,/dev/mapper/ubuntu--gnome--vg-root,4950740992,2921213952,2923085824,37.1,4950740992,1249525760,3918172160,331452416,7871954944,584634368,TP-LINK_BB28,True,-39,ssid,wpa2,70/70,Freebox-61A799,True,-85,ssid,wpa2,25/70,FreeWifi,False,-86,ssid,,24/70,FreeWifi_secure,True,-86,ssid,wpa2,24/70,FreeWifi_secure,True,-37,ssid,wpa2,70/70,freebox_NH,True,-37,ssid,wpa,70/70,FreeWifi,False,-39,ssid,,70/70,0.0,0.0,265,11.2,0.0,641,0.2239229679107666,62.9,25.8,0.0,0.0,4,0,4.9,0.0,1895,0.0
```

**Comment by tnibert** [comment]
So I poked around at this today and may have some clue as to why it is happening on my system but not yours.

In exports/glance_csv.py, I put a "print(stat)" in the update() method after "for stat in all_stats[i]:".  Everything progresses well until stat=
`{'cpu_times': pcputimes(user=700.83, system=79.13, children_user=265.39, children_system=29.37), 'cpu_percent': 24.7, 'nice': 0, 'pid': 14477, 'memory_percent': 10.403569919520097, 'ppid': 14424, 'status': 'S', 'username': 'tim', 'cmdline': ['/home/tim/.local/share/umake/ide/pycharm/jre64/bin/java', '-Xbootclasspath/a:/home/tim/.local/share/umake/ide/pycharm/lib/boot.jar', '-classpath', '/home/tim/.local/share/umake/ide/pycharm/lib/bootstrap.jar:/home/tim/.local/share/umake/ide/pycharm/lib/extensions.jar:/home/tim/.local/share/umake/ide/pycharm/lib/util.jar:/home/tim/.local/share/umake/ide/pycharm/lib/jdom.jar:/home/tim/.local/share/umake/ide/pycharm/lib/log4j.jar:/home/tim/.local/share/umake/ide/pycharm/lib/trove4j.jar:/home/tim/.local/share/umake/ide/pycharm/lib/jna.jar', '-Xms128m', '-Xmx750m', '-XX:ReservedCodeCacheSize=240m', '-XX:+UseConcMarkSweepGC', '-XX:SoftRefLRUPolicyMSPerMB=50', '-ea', '-Dsun.io.useCanonCaches=false', '-Djava.net.preferIPv4Stack=true', '-XX:+HeapDumpOnOutOfMemoryError', '-XX:-OmitStackTraceInFastThrow', '-Dawt.useSystemAAFontSettings=lcd', '-Dsun.java2d.renderer=sun.java2d.marlin.MarlinRenderingEngine', '-XX:ErrorFile=/home/tim/java_error_in_PYCHARM_%p.log', '-XX:HeapDumpPath=/home/tim/java_error_in_PYCHARM.hprof', '-Didea.paths.selector=PyCharmCE2017.2', '-Djb.vmOptionsFile=/home/tim/.local/share/umake/ide/pycharm/bin/pycharm64.vmoptions', '-Didea.platform.prefix=PyCharmCore', 'com.intellij.idea.Main'], 'gids': pgids(real=1000, effective=1000, saved=1000), 'io_counters': [23056384, 169537536, 23056384, 169537536, 1], 'num_threads': 49, 'name': 'java', 'memory_info': pmem(rss=859250688, vms=4774031360, shared=39833600, text=4096, lib=0, data=1017163776, dirty=0), 'time_since_update': 0.11886811256408691}
`

It then hits self.get_item_key(stat), where it attempts to assign the ret variable from a key named 'key'.  Since this key doesn't exist in the above stat, it raises exception and doesn't assign ret.  After this it calls isinstance(ret, list) and throws an exception because ret is not initialized.

I'll keep poking it with a stick later and see how that stat is gathered.

**Comment by tnibert** [comment]
Ok, if my flags are any indication, the exception is always raised on stat dicts of the form:
`{'cmdline': [], 'gids': pgids(real=0, effective=0, saved=0), 'memory_info': pmem(rss=0, vms=0, shared=0, text=0, lib=0, data=0, dirty=0), 'nice': 0, 'ppid': 2, 'memory_percent': 0.
0, 'name': 'kworker/1:2', 'pid': 28374, 'cpu_times': pcputimes(user=0.0, system=0.0, children_user=0.0, children_system=0.0), 'num_threads': 1, 'cpu_percent': 0.0, 'io_counters': [0, 0, 0, 0
, 0], 'username': 'root', 'status': '?', 'time_since_update': 1}`

My flags also seem to tell me these are from the fs plugin?

**Comment by balayanr** [comment]
Same error happening on a fresh install (v2.11.1 with psutil v5.4.2, Ubuntu 18.04).

`Exception in thread Thread-17:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3/dist-packages/glances/exports/glances_csv.py", line 83, in update
    plugin, self.get_item_key(stat), item) for item in stat)
  File "/usr/lib/python3/dist-packages/glances/exports/glances_csv.py", line 83, in <genexpr>
    plugin, self.get_item_key(stat), item) for item in stat)
  File "/usr/lib/python3/dist-packages/glances/exports/glances_export.py", line 119, in get_item_key
    if isinstance(ret, list):
UnboundLocalError: local variable 'ret' referenced before assignment
`

There is a clause in glances_export.get_item_key() that tries to get the key, but the except clause does not prevent the rest of the function from executing. Therefore, if the try clause fails to set up "ret" variable, but it's references after the try-except clause, you get an error. 

I added a print statement to the except clause and it looks like the app was calling get_item_key() on this item:
{'host': '192.168.11.1', 'port': 0, 'description': 'DefaultGateway', 'refresh': 30, 'timeout': 3, 'status': 0.002415, 'rtt_warning': None}
Seems like this "stat" should't really be in the export, or the way the program is handling item["key"] is strange.

Either way, as a temporary solution try adding a try-except in glances_csv.update(), in the section where it goes over plugins, seems to solve the problem. I just wrote "pass" in the except clause, you might want to log it or something.
